lib/libc/sys/caprevoke.2
.\" Copyright (c) 2020 Microsoft, Inc.
.\" All rights reserved.
.\"
.\" This documentation was written by
.\" Nathaniel Filardo <nfilardo@microsoft.com>
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHORS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHORS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\" $FreeBSD$
.\"
.Dd August 20, 2020
.Dt CAPREVOKE 2
.Os
.Sh NAME
.Nm caprevoke
.Nm caprevoke_shadow
.Nd interface to sweeping revocation of CHERI capabilities
.Sh LIBRARY
.Lb libc
.Sh SYNOPSIS
.In sys/caprevoke.h
.Ft int
.Fn caprevoke_shadow "int flags" "void * __capability arena" "void * __capability * shadow"
.Ft int
.Fn caprevoke "int flags" "caprevoke_epoch start" "struct caprevoke_stats *"
.Sh DESCRIPTION
The kernel exposes a CHERI capability revocation service,
a mechanism to revoke capabilities to regions of the address space.
Requests for revocation are made by setting bits
in the shadow bitmap and
invokving the
.Fn caprevoke
system call.
.Pp
The
.Fn caprevoke_shadow
system call is used to obtain access to segments of the shadow bitmap
and to global revocation metadata.
The
.Fa flags
argument governs the behavior of the call:
.Bl -tag
.It Bq CAPREVOKE_SHADOW_NOVMMAP
.Fa arena
should be a capability bearing the
.Dv VMMAP
software permission,
as returned by, for example,
.Xr mmap 2 ,
and it will store to its
.Fa shadow
argument a capability authorizing access to the shadow of the
.Fa arena .
A bit set in the region authorized by
.Fa shadow
corresponds to a request for revocation of capabilities lacking the
.Dv VMMAP
permission and whose
.Em base
points to the granule corresponding to that bit.
.It Bq CAPREVOKE_SHADOW_INFO_STRUCT
.Fa arena
is ignored and
.Fa shadow
will be set to a pointer to the global
.Dv caprevoke_info
structure,
which the kernel maintains with the current epoch counters.
.El
.Pp
The
.Fn caprevoke
system call
interfaces with this CHERI capability revocation service
to advance or synchronize with
the revocation state machine and revocation epoch clock.
Its
.Fa start
argument is used for relaxed synchronization
between multiple users of the revocation service.
If the epoch clock has advanced beyond the provided value,
the function will generally return immediately.
The
.Fa caprevoke_stats
argument is provided mostly for experimentation
and may be
.Dv NULL .
Its
.Fa flags
argument is,
for the moment,
mostly reserved for experimentation.
The sole value that may be of interest is
.Dv CAPREVOKE_IGNORE_START
which causes the
.Fa start
argument to be ignored
and the epoch counter
obtained by synchronizing with the revocation service
to be used instead.
Callers should always pass
.Dv CAPREVOKE_LAST_PASS
as well unless they are engaging in multi-pass experiments;
see the source.
.Sh EXAMPLES
.Pp
At allocator startup or upon first need, obtain a capability to the global
.Dv caprevoke_info
structure,
and store it to an allocator local variable:
.Bd -literal
    static struct caprevoke_info caprev_info;

    int res = caprevoke_shadow(CAPREVOKE_SHADOW_INFO_STRUCT, NULL, &info);
    assert(res == 0);
.Ed
.Pp
For a very simple allocator,
it suffices to collect
.Fn free
memory into quarantine and then revoke it all at once:
.Bd -literal
    caprevoke_shadow(CAPREVOKE_SHADOW_NOVMMAP, arena, &shadow);
    /* elided: set bits in shadow corresponding to quarantine */

    atomic_thread_fence(memory_order_acq_rel);
    caprevoke_epoch epoch_start = info->epoch_enqueue;

    while(!caprevoke_epoch_clears(info->epoch_dequeue, epoch_start)) {
      caprevoke(CAPREVOKE_LAST_PASS, epoch_start, NULL);
    }

    /* elided: clear bits in shadow corresponding to quarantine */
    atomic_thread_fence(memory_order_rel);

    /* It is now safe to reuse quarantined memory */
.Ed
.Pp
The release fence may be merged into a subsequent lock release, but the
acquire-release fence is requisite in almost all cases.
.Pp
For more complex allocators, several steps are necessary.
.Pp
Whenever allocating a new arena using
.Xr mmap 2
also obtain access to the shadow space:
.Bd -literal
    struct arena_meta *meta = // ...
    meta->ptr = mmap(/* ... */);
 
    int res = caprevoke_shadow(CAPREVOKE_SHADOW_NOVMMAP, meta->ptr,
                               &meta->caprev_sh);
    assert(res == 0);
.Ed
.Pp
When a chunk of quarantine fills,
label it with its
.Em enqueue
epoch counter:
.Bd -literal
    // elided: fill out the shadow bitmap prior to here
    atomic_thread_fence(memory_order_acq_rel); // publish shadow
    chunk->caprev_enq = info->epoch_enqueue; // label this chunk
.Ed
.Pp
When revocation becomes necessary,
loop until the
.Em dequeue
epoch counter clears the eldest chunk,
and then return all usable chunks to service:
.Bd -literal
    while(!caprevoke_epoch_clears(info->epoch_dequeue, chunk->caprev_enq)) {
      caprevoke(CAPREVOKE_LAST_PASS, chunk->caprev_enq, NULL);
    }
    more = chunk->next;
    /* This chunk is certainly reusable now */
    reusable(chunk);

    /* And others may be as well */
    for(chunk = more; 
        chunk && caprev_epoch_clears(info->epoch_dequeue, chunk->caprev_enq);
        chunk = chunk->next) {
      reusable(chunk);
    }
.Ed
.Pp
The
.Fn reusable
function in this example is responsible for clearing shadow bits.
The allocator as a whole,
but not necessarily the
.Fn reusable
function itself,
must, as before,
execute a release fence
before actually reusing memory.
.Sh ERRORS
.Fn caprevoke_info
fails with
.Bl -tag -width Er
.It Bq Er EPERM
if the
.Fa arena
for
.DV CAPREVOKE_SHADOW_MEM
does not cary the
.Dv VMMAP
permission.
.It Bq Er EINVAL
if the
.Fa arena
for
.Dv CAPREVOKE_SHADOW_MEM
is not sufficiently aligned and sized.
.It Bq Er EINVAL
if the
.Fa shadow
parameter for either
.Dv CAPREVOKE_SHADOW_MEM
or
.Dv CAPREVOKE_SHADOW_INFO
points to an invalid address.
.El
.Pp
.Fn caprevoke
will return 0 if the
.Fa start
epoch has been cleared by time of return;
otherwise, it reutrns
.Bl -tag -width Er
.It Bq Er EAGAIN
if the
.Fa start
epoch is presently open or in the future
.It Bq Er EINVAL
if invalid
.Fa flags
have been specified
.It Bq Er EINVAL
if
.Fa start
does not describe an epoch and
.Dv CAPREVOKE_START_NOW
was not given in
.Fa flags .
.El


lib/libcheri_caprevoke/caprevoke.c
/*
 * Revocation bitmap manipulation operations.
 */

// XXX This can't possibly be right.  Should vmparam.h be checking for
// something else?
#ifndef CPU_CHERI
#define CPU_CHERI
#endif

#include <assert.h>
#include <inttypes.h>
#include <stdbool.h>
#include <stdatomic.h>
#include <stddef.h>
#include <stdint.h>

#include <cheri/cheric.h>

#include <sys/param.h>
#include <sys/caprevoke.h>
#include <machine/cherireg.h>
#include <machine/pte.h>
#include <machine/vmparam.h>
#include "caprevoke.h"

static ptrdiff_t caprev_shadow_nomap_first_word_offset(vaddr_t base);
static ptrdiff_t caprev_shadow_nomap_last_word_offset(vaddr_t base, size_t len);
static uint64_t  caprev_shadow_nomap_first_word_mask(vaddr_t base, size_t len);
static uint64_t  caprev_shadow_nomap_last_word_mask(vaddr_t base, size_t len);

  /*
   * Given a capability to an object, derive the offset relative to the base of
   * the fine-grained bitmap where the first word would be located.
   */

static ptrdiff_t
caprev_shadow_nomap_first_word_offset(vaddr_t base)
{
  return (ptrdiff_t)(base / VM_CAPREVOKE_GSZ_MEM_NOMAP
                          / 8
                          / sizeof(uint64_t)) * sizeof(uint64_t);
}

static ptrdiff_t
caprev_shadow_nomap_last_word_offset(vaddr_t base, size_t len)
{
  if (len == 0)
    return caprev_shadow_nomap_first_word_offset(base);

  return (ptrdiff_t)((base + len - 1)
                          / VM_CAPREVOKE_GSZ_MEM_NOMAP
                          / 8
                          / sizeof(uint64_t)) * sizeof(uint64_t);
}

  /*
   * Derive the mask to be or'd in to the first word to mark the object's
   * bottom granules as revoked.
   */

static uint64_t
caprev_shadow_nomap_first_word_mask(vaddr_t base, size_t len)
{
  uint64_t res;

  /* What's the least significant bit's position within the word? */
  int lsb = (base / VM_CAPREVOKE_GSZ_MEM_NOMAP) % (8 * sizeof(uint64_t));

  if (caprev_shadow_nomap_first_word_offset(base) ==
      caprev_shadow_nomap_last_word_offset(base, len)) {
    /* The object occupies only some bits in the first word */

    int setwidth = len / VM_CAPREVOKE_GSZ_MEM_NOMAP;

    if (lsb + setwidth == 64) {
      /* Object fills this word completely, but does not spill to the next */

      res = ~(((uint64_t)1 << lsb) - 1);
    } else {
      res = (   ~(((uint64_t)1 << (lsb + setwidth)) - 1)
             ^ (~(((uint64_t)1 << lsb             ) - 1)));
    }
  } else {
    /* The object runs off the end of this word. */

    res = ~(((uint64_t)1 << lsb) - 1);
  }

#if __BYTE_ORDER__ == __ORDER_BIG_ENDIAN__
  return __builtin_bswap64(res);
#elif __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__
  return res;
#else
# error
#endif
}

static uint64_t
caprev_shadow_nomap_last_word_mask(vaddr_t base, size_t len)
{
  uint64_t res;

  if (caprev_shadow_nomap_first_word_offset(base) ==
      caprev_shadow_nomap_last_word_offset(base, len)) {

    /*
     * There are no more bits to set that haven't been taken care of by the the
     * first word, so just return 0.
     */

    return 0;
  }

  int msb = ((base + len - 1) / VM_CAPREVOKE_GSZ_MEM_NOMAP) % (8 * sizeof(uint64_t));

  if (msb == 63) {
    /* This object runs to the end of its last word */

    return ~(uint64_t)0;
  }

  res = ((uint64_t)1 << (msb+1)) - 1;

#if __BYTE_ORDER__ == __ORDER_BIG_ENDIAN__
  return __builtin_bswap64(res);
#elif __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__
  return res;
#else
# error
#endif
}

  /*
   * Exposed for testing, but hopefully inlined into the actual update
   * functions below.
   *
   * XXX Slightly less polluting of the external namespace than having all four
   * functions available, maybe?
   */

void
caprev_shadow_nomap_offsets(vaddr_t ob, size_t len, ptrdiff_t *fwo, ptrdiff_t *lwo)
{
  *fwo = caprev_shadow_nomap_first_word_offset(ob);
  *lwo = caprev_shadow_nomap_last_word_offset(ob, len);
}

void
caprev_shadow_nomap_masks(vaddr_t ob, size_t len, uint64_t *fwm, uint64_t *lwm)
{
  *fwm = caprev_shadow_nomap_first_word_mask(ob, len);
  *lwm = caprev_shadow_nomap_last_word_mask(ob, len);
}

  /*
   * A helper function for some things both setting and clearing shadow masks
   * need up front.
   */

static inline void
caprev_shadow_nomap_common_pfx(uint64_t * __capability sb, void * __capability obj,
             vaddr_t *ob, size_t *len, ptrdiff_t *fwo, ptrdiff_t *lwo,
             uint64_t * __capability *fw
            )
{
  *ob = (__cheri_addr vaddr_t)obj;
  *len = cheri_getlen(obj);

  caprev_shadow_nomap_offsets(*ob, *len, fwo, lwo);

  *fw = cheri_setaddress(sb, VM_CAPREVOKE_BM_MEM_NOMAP + *fwo);
}

  /*
   * Mark an object for revocation using the indicated capability to the
   * correct fragment of the shadow bitmap.
   *
   * The "privileged" object capability is expected to point to the full,
   * actual bounds of the allocation and may (or may not) be subject to
   * revocation; this capability determines which bits of the bitmap are in
   * scope for changes.  The "user" object capability is expected to be a
   * subset of the privileged object capability and, importantly, must be
   * subject to revocation; this capability is used to guard against
   * double-frees.  The two are provided separately to remove from the
   * allocator the need to construct the priveleged capability while
   * respecting (concurrently mutable!) taggedness of the user capability.
   *
   * The masks here are additive: they're going to be OR'd with the shadow.
   *
   * Returns 0 on success or 1 if the calling free() should skip this
   * capability, as it has either become de-tagged or is at least partially
   * already marked in the bitmap (as might happen, for example, with a
   * racing second call to free on the same pointer).
   */

int
caprev_shadow_nomap_set(uint64_t * __capability sb,
                        void * __capability priv_obj,
                        void * __capability user_obj)
{
  vaddr_t ob;
  size_t len;
  ptrdiff_t fwo, lwo;
  uint64_t * __capability fw;

  caprev_shadow_nomap_common_pfx(sb, priv_obj, &ob, &len, &fwo, &lwo, &fw);

  uint64_t fwm = caprev_shadow_nomap_first_word_mask(ob, len);

  uint64_t asmres = 0;

  /*
   * The first word is tricky to handle, since we use it to gate frees,
   * ensuring that double-frees are idempotent.  There are three reasons we
   * might bail out:
   *
   *   - another thread has quarantined this object (bits already set)
   *
   *   - this object is detagged
   *
   *   - this object has been revoked while we held a pointer to it in
   *   free() (and so its permissions have been stripped; see cheri_revoke)
   *
   * We have to check the permissions inside the ll/sc we use to update the
   * shadow bitmask so that we interlock against the stop-the-world (and
   * maybe read-side) phases of revocation.
   *
   * It's possible the SC fails for neither of the above reasons, of course,
   * not the least of which is that a concurrent thread might be mutating
   * the shadow.  In that case, just go around again.
   *
   * We're going to do this dance with some inline assembler (sigh), because
   * it's a little much to expect C to understand!
   */
#ifdef __mips__

  // t0: the local, manipulated copy of the shadow
  // t1: temp value

  // XXX Why is someone inserting nops in my carefully-written assembler? :(

  __asm__ __volatile__ (
      "1:\n\t"
      "clld $t0, %[fw]\n\t"             // Load linked the first word

      "and $t1, $t0, %[fwm]\n\t"        // Jump out if shadow set
      "bnez $t1, 2f\n\t"
      "nop\n\t"                         // [delay]

      "cbtu %[obj], 2f\n\t"             // Jump out if object detagged
	// no delay slot OK

      "cgetperm $t1, %[obj]\n\t"        // Jump out if zero perms
      "beqz $t1, 2f\n\t"
      "nop\n\t"                         // [delay]

      "or $t0, $t0, %[fwm]\n\t"         // bitwise or in the mask [delay]

      "cscd %[asmres], $t0, %[fw]\n\t"  // SC the updated mask
      "bnez %[asmres], 2f\n\t"          // Jump out on success
      "nop\n\t"                         // [delay]

      "j 1b\n\t"                        // Go again if SC failed
      "nop\n\t"

      "2:\n\t"
    : // output operands
      [asmres] "+r" (asmres)
    : // input operands
      [obj]    "C"  (user_obj),
      [fw]     "C"  (fw),
      [fwm]    "r"  (fwm)
     : // clobbers
       "t0", "t1", "memory"
  );
#else
#  error Unable to mark bitmaps on this architecture
#endif

  /*
   * In the "Jump out" cases above, %[asmres] will hold the value 0, either
   * from initialization or because a cscd instruction was unsuccessful on a
   * prior iteration.  Only if we don't jump out and successfully exit the loop
   * will asmres be 1.
   */
  if (asmres == 0) {
    return 1;
  }

  if (lwo != fwo) {

    /*
     * If this object straddles a word boundary, the first and last word offets
     * will not be equal; the last word might also need special handling.
     */

    uint64_t * __capability lw = cheri_setaddress(sb,
                                    VM_CAPREVOKE_BM_MEM_NOMAP + lwo);
    uint64_t lwm = caprev_shadow_nomap_last_word_mask(ob, len);

    /*
     * We might overlap a concurrent object's attempt to mutate bits in the
     * last word, so do the OR atomically unless we're setting the whole word,
     * in which case, the write cannot possibly race.
     */

    if (lwm != ~(uint64_t)0) {
      atomic_fetch_or_explicit((_Atomic(uint64_t) * __capability)lw, lwm,
                               memory_order_relaxed);
    } else {
      *lw = ~(uint64_t)0;
    }

    /*
     * Any words between the first and last we can just set all at once;
     * there's nothing else to see here.
     */

    ptrdiff_t wo = fwo + sizeof(uint64_t);
    uint64_t * __capability sbo = fw + 1;
    for (; wo < lwo; wo += sizeof(uint64_t), sbo++) {
      *sbo = ~(uint64_t)0;
    }
  }

  /*
   * There is no need to store-barrier or any such here; we can enforce store
   * order before grabbing the epoch clock for this *chunk* of the quarantine.
   * By ensuring that we make all cores aware of our bits before looking at the
   * clock, we guarantee that the revocation epoch(s) upon which we are waiting
   * will be done with at least our stores visible.
   */

  return 0;
}

  /*
   * Clear an object from the shadow bitmap.
   *
   * The masks here are subtractive: they're going to be AND'd with the shadow.
   */

void
caprev_shadow_nomap_clear(uint64_t * __capability sb, void * __capability obj)
{
  vaddr_t ob;
  size_t len;
  ptrdiff_t fwo, lwo;
  uint64_t * __capability fw;

  caprev_shadow_nomap_common_pfx(sb, obj, &ob, &len, &fwo, &lwo, &fw);

  uint64_t fwm = ~caprev_shadow_nomap_first_word_mask(ob, len);

  /*
   * The first word must be handled atomically, so that it interlocks
   * with the atomic manipulation in caprev_shadow_nomap_set.
   */
  atomic_fetch_and_explicit((_Atomic(uint64_t) * __capability)fw, fwm,
                            memory_order_relaxed);

  if (lwo != fwo) {

    uint64_t * __capability lw = cheri_setaddress(sb,
                                    VM_CAPREVOKE_BM_MEM_NOMAP + lwo);
    uint64_t lwm = ~caprev_shadow_nomap_last_word_mask(ob, len);

    /*
     * The last word may also need to be handled atomically.
     */

    if (lwm != 0) {
      atomic_fetch_and_explicit((_Atomic(uint64_t) * __capability)lw, lwm,
                                memory_order_relaxed);
    } else {
      *lw = 0;
    }

    /*
     * Any words between the first and last we can just clear all at once.
     */

    ptrdiff_t wo = fwo + sizeof(uint64_t);
    uint64_t * __capability sbo = fw + 1;
    for (; wo < lwo; wo += sizeof(uint64_t), sbo++) {
      *sbo = (uint64_t)0;
    }
  }

  /*
   * There is no need to store-barrier or any such here either.  It suffices to
   * store-barrier at the end of whatever set of objects is being cleared from
   * the shadow, before those objects may be released to the application.
   */
}

/*
 * The "raw" interfaces presume that all interlocking has already been
 * managed by the caller and so contain no atomics or such!
 */

void
caprev_shadow_nomap_set_raw(uint64_t * __capability sb,
                            vaddr_t heap_start, vaddr_t heap_end)
{
  ptrdiff_t fwo, lwo;
  uint64_t fwm;
  uint64_t * __capability fw;
  size_t len = heap_end - heap_start;

  caprev_shadow_nomap_offsets(heap_start, len, &fwo, &lwo);

  fw = cheri_setaddress(sb, VM_CAPREVOKE_BM_MEM_NOMAP + fwo);
  *fw |= caprev_shadow_nomap_first_word_mask(heap_start, len);

  if (lwo != fwo) {
    uint64_t lwm;
    uint64_t * __capability w = fw + 1;
    ptrdiff_t wo = fwo + sizeof(uint64_t);
    w = fw + 1;
    for (; wo < lwo; wo += sizeof(uint64_t), w++) {
		*w = ~(uint64_t)0;
	}

    w = cheri_setaddress(sb, VM_CAPREVOKE_BM_MEM_NOMAP + lwo);
    *w |= caprev_shadow_nomap_last_word_mask(heap_start, len);
  }
}

void
caprev_shadow_nomap_clear_raw(uint64_t * __capability sb,
                              vaddr_t heap_start, vaddr_t heap_end)
{
  ptrdiff_t fwo, lwo;
  uint64_t fwm;
  uint64_t * __capability fw;
  size_t len = heap_end - heap_start;

  caprev_shadow_nomap_offsets(heap_start, len, &fwo, &lwo);

  fw = cheri_setaddress(sb, VM_CAPREVOKE_BM_MEM_NOMAP + fwo);
  *fw &= ~caprev_shadow_nomap_first_word_mask(heap_start, len);

  if (lwo != fwo) {
    uint64_t lwm;
    uint64_t * __capability w = fw + 1;
    ptrdiff_t wo = fwo + sizeof(uint64_t);
    w = fw + 1;
    for (; wo < lwo; wo += sizeof(uint64_t), w++) {
		*w = 0;
	}

    w = cheri_setaddress(sb, VM_CAPREVOKE_BM_MEM_NOMAP + lwo);
    *w &= ~caprev_shadow_nomap_last_word_mask(heap_start, len);
  }
}
